error_handler.py

```diff
@@ -0,0 +1,114 @@
+
+
+def get_full_class_name(obj):
+    module = obj.__class__.__module__
+    if module is None or module == str.__class__.__module__:
+        return obj.__class__.__name__
+    return module + '.' + obj.__class__.__name__
+
+class NonBugError(Exception):
+    """When this is raised instead of a normal Exception, on_command_error() will not attach a traceback or github
+    link. """
+    pass
+
+# completely overkill error handler shamelessly ripped from MediaForge, a bot i wrote
+class ErrorHandlerCog(commands.Cog):
+    def __init__(self, bot):
+        self.bot: commands.Bot = bot
+        self.antispambucket = {}
+
+    @commands.Cog.listener()
+    async def on_command_error(self, ctx: commands.Context, commanderror: commands.CommandError):
+        async def dmauthor(*args, **kwargs):
+            try:
+                return await ctx.reply(*args, **kwargs)
+            except discord.Forbidden:
+                logger.info(f"Reply to {ctx.message.id} and dm to {ctx.author.id} failed. Aborting.")
+
+        async def reply(msg, file=None, embed=None):
+            if ctx.interaction:
+                if ctx.interaction.response.is_done():
+                    return await ctx.interaction.edit_original_response(content=msg,
+                                                                        attachments=[file] if file else None,
+                                                                        embed=embed)
+                else:
+                    return await ctx.reply(msg, file=file, embed=embed)
+            else:
+                try:
+                    if ctx.guild and not ctx.channel.permissions_for(ctx.me).send_messages:
+                        logger.debug(f"No permissions to reply to {ctx.message.id}, trying to DM author.")
+                        return await dmauthor(msg, file=file, embed=embed)
+                    return await ctx.reply(msg, file=file, embed=embed)
+                except discord.Forbidden:
+                    logger.debug(f"Forbidden to reply to {ctx.message.id}, trying to DM author")
+                    return await dmauthor(msg, file=file, embed=embed)
+
+        async def logandreply(message):
+            if ctx.guild:
+                ch = f"channel #{ctx.channel.name} ({ctx.channel.id}) in server {ctx.guild} ({ctx.guild.id})"
+            else:
+                ch = "DMs"
+            logger.info(f"Command '{ctx.message.content}' by "
+                        f"@{ctx.message.author.name}#{ctx.message.author.discriminator} ({ctx.message.author.id}) "
+                        f"in {ch} failed due to {message}.")
+            await reply(message)
+
+        errorstring = str(commanderror)
+        if isinstance(commanderror, discord.Forbidden):
+            await dmauthor(f"I don't have permissions to send messages in that channel.")
+            logger.info(commanderror)
+        if isinstance(commanderror, discord.ext.commands.errors.CommandNotFound):
```

Comment:
can we replace this with match? the isinstance spam is hard to look at
and also `import discord.ext.commands.errors as boterror` would be nice

Reply:
I thought about this. I can't find clear info on whether or not Python match statements return values or not (I'm guessing they don't), but it'd be nice if there were a way to extract out the repeated `await logandreply(err)` calls as well. I got distracted by not being quite able to figure out how to make it all work.

Reply:
does python match work with isinstance? I thought it didn't

Reply:
Seemingly, but I'm not 100% clear how exactly that all works. https://peps.python.org/pep-0636/#adding-a-ui-matching-objects

Reply:
> does python match work with isinstance? I thought it didn't

matches primary use is replacing isinstance with just raw type matching, ~~copied off~~ similar to rust

Reply:
https://www.geeksforgeeks.org/python-match-case-statement/ clarifies a bit:

> We can also use [Python classes](https://www.geeksforgeeks.org/python-classes-and-objects/) to match using the Python match statements. This will make the code very much clean,  neat, and more importantly easily readable.
>
> The first thought after seeing the match statement would be that the Programmer and Person inside the match statement would create an instance of the respective classes, but it is not the case. It will check for the instance with the given values. So the case Programmer(“Om”, language = “Python”, framework = “Django”)  will not create an instance but will check that the object passed whose name is instance is a  real instance of the Programmer class, then it will check for the name that is  “Om”, language that is “Python” and then framework that is “Django”. This way we can check for the values in the instance of any class.

Check the section "Use of Python Class in match statement" for more, but the fact that it's using construction syntax for an attribute list is at least a way of explaining what's going on, even if it's a bit surprising.

The full rules, given at https://docs.python.org/3/reference/compound_stmts.html#class-patterns , are rather involved.

Reply:
Already updated the code to use a match statement